File: adaptive-kernel/capabilities/safe_shell.jl
# capabilities/safe_shell.jl - Restricted shell execution
# SECURITY HARDENED: Blocked dangerous characters, null bytes, newlines, globbing
# PATH TRAVERSAL FIX: Added canonicalization, whitelist directories, explicit errors
# SHELL INJECTION FIX: Enhanced with comprehensive pattern matching

module SafeShell

export meta, execute

# ============================================================================
# SECURITY CONFIGURATION - FAIL CLOSED
# ============================================================================

# Whitelist of allowed commands (for security)
# SECURITY HARDENED: Principle of Least Privilege - only safe output commands
# REMOVED: cat, head, tail, wc (file-reading commands enable data exfiltration)
# REMOVED: md5sum, sha256sum (could be used to verify exfiltrated data hashes)
const WHITELIST = Set([
    # Safe output-only commands
    "echo",
    # System info (minimal, non-disclosing)
    "date", "hostname"
])

# SECURITY: Allowed base directories for file operations
# PATH TRAVERSAL FIX: Restrict operations to safe directories only
const ALLOWED_DIRS = [
    "/tmp",
    "/var/tmp",
    "/home/user/projectx"  # Project workspace
]

# Shell metacharacters that could enable injection
# SECURITY: Enhanced with more dangerous patterns
const DANGEROUS_CHARS = r"[;&|`\$(){}<>\\]|&&|\|\||\s+>"

# SECURITY: Additional dangerous patterns to block
# SECURITY HARDENED: Extended to cover all sensitive system directories
const BLOCKED_PATTERNS = [
    r"\x00",  # Null byte injection
    r"\n",    # Newline injection  
    r"\r",    # Carriage return injection
    r"\s+-",  # Flag injection (e.g., "ls -la")
    r"\*",    # Glob expansion
    r"\?",    # Glob wildcard
    r"\[.*\]",  # Glob bracket
    r"\~",    # Tilde expansion
    r"\.\.",  # Directory traversal
    # Sensitive system directories - BLOCK ALL ACCESS
    r"^/etc",    # System configuration
    r"^/proc",   # Process information
    r"^/sys",    # System kernel info
    r"^/root",   # Root home directory
    r"^/var",    # Variable data (logs, caches, etc)
    r"^/opt",    # Optional/third-party software
    r"^/srv",    # Service data
    r"^/mnt",    # Mount points
    r"^/boot",   # Boot files
    r"^/media",  # Removable media
    r"^/dev",    # Device files
    r"^/run",    # Runtime data
    r"^/lost\+found",  # Filesystem recovery
]

# SECURITY FIX: Shell injection specific patterns
const SHELL_INJECTION_PATTERNS = [
    r"\$\([^)]+\)",           # Command substitution $(...)
    r"`[^`]+`",                 # Backtick command substitution
    r"\|\s*bash",              # Pipe to bash
    r"\|\s*sh",                # Pipe to sh
    r"&&\s+",                  # Command chaining
    r"\|\s*",                  # Pipe commands
    r";#",                      # Comment injection
    r"\s>\s*/tmp",             # Output redirect to tmp
    r"\s>>\s*/tmp",            # Output append to tmp
    r"\s\d?>\s*",             # Any output redirect
    r";\s*rm\s",              # Semicolon + rm
    r";\s*wget\s",            # Semicolon + wget
    r";\s*curl\s",            # Semicolon + curl
    r";\s*nc\s",              # Semicolon + netcat
    r";\s*nmap\s",            # Semicolon + nmap
    r"\|\s*nc\s",             # Pipe + netcat
    r"\|\s*nmap\s",           # Pipe + nmap
    r"\bexec\s",              # exec command
    r"\beval\s",              # eval command
    r"\bsource\s",            # source command
    r"\.\s*/",                # Execute from path
]

# Maximum command length to prevent buffer overflow attacks
const MAX_COMMAND_LENGTH = 256

# SHELL INJECTION FIX: Validate input is purely alphanumeric
function validate_no_injection(command::String)::Tuple{Bool, String}
    # Check for command substitution
    for pattern in SHELL_INJECTION_PATTERNS
        if occursin(pattern, command)
            return (false, "Shell injection pattern detected: $(pattern)")
        end
    end
    
    # Check for dangerous characters
    if occursin(DANGEROUS_CHARS, command)
        return (false, "Shell metacharacters detected")
    end
    
    # Ensure command only contains safe characters
    if !occursin(r"^[a-zA-Z0-9_\-\s]+$", command)
        return (false, "Command contains invalid characters")
    end
    
    return (true, "OK")
end

function meta()::Dict{String, Any}
    return Dict(
        "id" => "safe_shell",
        "name" => "Safe Shell Execute",
        "description" => "Execute a whitelisted command in a restricted shell",
        "inputs" => Dict("command" => "string"),
        "outputs" => Dict("stdout" => "string", "exit_code" => "int", "success" => "bool"),
        "cost" => 0.1,
        "risk" => "high",
        "reversible" => false
    )
end

"""
    sanitize_argument - Sanitize individual command arguments
    SECURITY: Prevents injection through arguments
    
    PATH TRAVERSAL FIX: Returns (success, sanitized_value, error_message) tuple
    instead of throwing exceptions. This ensures proper error handling.
"""
function sanitize_argument(arg::String)::Tuple{Bool, String, String}
    # Check for blocked patterns - return error instead of throwing
    for pattern in BLOCKED_PATTERNS
        if occursin(pattern, arg)
            # PATH TRAVERSAL FIX: Return explicit error instead of throwing
            return (false, "", "SECURITY: Blocked dangerous pattern detected: $(pattern)")
        end
    end
    
    # SHELL INJECTION FIX: Additional validation for arguments
    valid, error_msg = validate_no_injection(arg)
    if !valid
        return (false, "", "SECURITY: $error_msg")
    end
    
    # Remove any remaining special characters
    sanitized = replace(arg, r"[^a-zA-Z0-9_\-./]" => "")
    return (true, sanitized, "")
end

# Allow SubString input as well
sanitize_argument(arg::SubString{String}) = sanitize_argument(String(arg))

"""
    validate_command - Comprehensive command validation
    SECURITY: Multiple layers of validation
    
    PATH TRAVERSAL FIX: Now handles tuple return from sanitize_argument
    and adds path canonicalization check.
"""

"""
    canonicalize_path - Resolve path to canonical form
    PATH TRAVERSAL FIX: Resolves symlinks and .. components to prevent bypass attacks
"""
function canonicalize_path(path::String)::Tuple{Bool, String, String}
    # Skip empty paths
    if isempty(path)
        return (true, "", "")
    end
    
    # Skip if path doesn't exist (will be caught by command execution)
    # But check for ALL suspicious patterns first
    # SECURITY FIX: Extended to cover all sensitive system directories
    if occursin(r"^/etc|^/proc|^/sys|^/root|^/var|^/opt|^/srv|^/mnt|^/boot|^/media|^/dev|^/run", path)
        return (false, "", "Access to sensitive system directory denied")
    end
    
    # Try to canonicalize if file exists - handle edge cases safely
    if isfile(path) || isdir(path)
        try
            canonical = realpath(path)
            # Check if canonical path is within allowed directories
            allowed = false
            for allowed_dir in ALLOWED_DIRS
                # Skip if allowed_dir doesn't exist
                if isdir(allowed_dir) && startswith(canonical, allowed_dir)
                    allowed = true
                    break
                end
            end
            if !allowed
                return (false, "", "Path resolved to non-allowed directory: $canonical")
            end
            return (true, canonical, "")
        catch e
            return (false, "", "Failed to canonicalize path: $(string(e))")
        end
    end
    
    # For non-existent paths, validate the path components safely
    # SECURITY FIX: Block ALL paths outside allowed directories, no warnings
    if startswith(path, "/")
        # Absolute path - check if it starts with any allowed directory
        for allowed_dir in ALLOWED_DIRS
            # Only check existing allowed directories
            if isdir(allowed_dir) && startswith(path, allowed_dir)
                return (true, path, "")
            end
        end
        # Not in allowed directory - BLOCK it (no warning, block directly)
        return (false, "", "Path not in allowed directories: $path")
    end
    
    return (true, path, "")
end

function validate_command(command::String)::Tuple{Bool, String}
    # Check 1: Length limit
    if length(command) > MAX_COMMAND_LENGTH
        return (false, "Command exceeds maximum length of $MAX_COMMAND_LENGTH")
    end
    
    # Check 2: Null byte
    if occursin("\x00", command)
        return (false, "Null byte injection detected")
    end
    
    # Check 3: Newline injection
    if occursin("\n", command) || occursin("\r", command)
        return (false, "Newline/carriage return injection detected")
    end
    
    # SHELL INJECTION FIX: Enhanced shell metacharacter check
    # Check 4: Shell metacharacters AND injection patterns
    if occursin(DANGEROUS_CHARS, command)
        return (false, "Shell metacharacters detected")
    end
    
    # Check 4b: Shell injection patterns (comprehensive)
    valid, error_msg = validate_no_injection(command)
    if !valid
        return (false, error_msg)
    end
    
    # Check 5: Get command base
    parts = split(command)
    if isempty(parts)
        return (false, "Empty command")
    end
    cmd_base = parts[1]
    
    # Check 6: Whitelist
    if cmd_base ∉ WHITELIST
        return (false, "Command '$cmd_base' not in whitelist")
    end
    
    # Check 7: Argument sanitization with path traversal detection
    # PATH TRAVERSAL FIX: Now handles tuple return and canonicalizes paths
    if length(parts) > 1
        for i in 2:length(parts)
            # Use new tuple return from sanitize_argument
            success, sanitized, error_msg = sanitize_argument(parts[i])
            
            if !success
                return (false, error_msg)
            end
            
            if isempty(sanitized) && !isempty(parts[i])
                return (false, "Dangerous argument detected: $(parts[i])")
            end
            
            # PATH TRAVERSAL FIX: Also canonicalize and check the path
            if !isempty(sanitized) && (startswith(sanitized, "/") || occursin("/", sanitized))
                canon_success, canonical_path, canon_error = canonicalize_path(sanitized)
                if !canon_success
                    return (false, "Path validation failed: $canon_error")
                end
            end
        end
    end
    
    return (true, "OK")
end

function execute(params::Dict)::Dict{String, Any}
    command = get(params, "command", "echo safe")
    
    # SECURITY: Comprehensive validation
    valid, reason = validate_command(command)
    if !valid
        # PATH TRAVERSAL FIX: Explicit error message instead of empty string
        return Dict(
            "success" => false,
            "effect" => "Command validation failed: $reason",
            "actual_confidence" => 0.0,
            "energy_cost" => 0.0,
            "data" => Dict(
                "stdout" => "ERROR: $reason",  # FIX: Explicit error in stdout
                "exit_code" => 127,
                "reason" => reason
            )
        )
    end
    
    try
        # Execute using safer method - avoid shell interpretation
        # Split command and sanitize arguments
        parts = split(command)
        cmd = parts[1]
        args = String[]
        
        if length(parts) > 1
            for i in 2:length(parts)
                # PATH TRAVERSAL FIX: Handle tuple return from sanitize_argument
                success, sanitized, error_msg = sanitize_argument(parts[i])
                if success && !isempty(sanitized)
                    push!(args, sanitized)
                elseif !success
                    # Security error during execution - should not happen if validation passed
                    return Dict(
                        "success" => false,
                        "effect" => "Security validation failed: $error_msg",
                        "actual_confidence" => 0.0,
                        "energy_cost" => 0.0,
                        "data" => Dict(
                            "stdout" => "ERROR: $error_msg",  # FIX: Explicit error
                            "exit_code" => 127,
                            "reason" => error_msg
                        )
                    )
                end
            end
        end
        
        # Execute command directly without shell interpretation
        # Use backtick syntax for safe command execution (safe against injection)
        cmd_expr = `$(cmd) $(args...)`
        output = read(cmd_expr, String)
        
        return Dict(
            "success" => true,
            "effect" => "Command executed: $cmd",
            "actual_confidence" => 0.85,
            "energy_cost" => 0.1,
            "data" => Dict(
                "stdout" => output,
                "exit_code" => 0,
                "command" => command
            )
        )
    catch e
        # PATH TRAVERSAL FIX: Explicit error message instead of empty string
        error_detail = string(e)
        return Dict(
            "success" => false,
            "effect" => "Execution failed: $error_detail",
            "actual_confidence" => 0.0,
            "energy_cost" => 0.0,
            "data" => Dict(
                "stdout" => "ERROR: $error_detail",  # FIX: Explicit error in stdout
                "exit_code" => 1,
                "reason" => error_detail
            )
        )
    end
end

end  # module

File: itheris-daemon/src/safe_shell.rs
//! # Safe Shell - Secure Command Execution with Jailing
//!
//! This module provides restricted shell execution with:
//! - Command whitelist (only safe commands allowed)
//! - Directory restrictions (prevent path traversal)
//! - Shell injection prevention
//! - Output size limits
//! - Execution timeout
//!
//! ## Security Properties
//!
//! - **Whitelist Only**: Only explicitly allowed commands
//! - **Directory Jailing**: Restricted to safe directories
//! - **Injection Prevention**: Blocks shell metacharacters
//! - **Size Limits**: Prevents buffer overflow
//! - **Timeout**: Prevents hanging commands

use regex::Regex;
use serde::{Deserialize, Serialize};
use std::collections::HashSet;
use std::path::Path;
use std::process::{Command, Stdio};
use std::time::{Duration, Instant};
use thiserror::Error;

/// Safe shell errors
#[derive(Error, Debug)]
pub enum SafeShellError {
    #[error("Command not allowed: {0}")]
    CommandNotAllowed(String),

    #[error("Directory not allowed: {0}")]
    DirectoryNotAllowed(String),

    #[error("Shell injection detected: {0}")]
    InjectionDetected(String),

    #[error("Output too large: {0}")]
    OutputTooLarge(String),

    #[error("Execution timeout: {0}")]
    Timeout(String),

    #[error("Execution failed: {0}")]
    ExecutionFailed(String),
}

/// Shell execution result
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ShellResult {
    pub success: bool,
    pub stdout: String,
    pub stderr: String,
    pub exit_code: i32,
    pub duration_ms: u64,
}

/// Shell configuration
#[derive(Clone)]
pub struct ShellConfig {
    /// Allowed directories
    pub allowed_dirs: Vec<String>,
    /// Allowed commands (whitelist)
    pub allowed_commands: HashSet<String>,
    /// Maximum output size in bytes
    pub max_output_size: usize,
    /// Execution timeout in seconds
    pub timeout_secs: u64,
    /// Maximum command length
    pub max_command_length: usize,
}

impl Default for ShellConfig {
    fn default() -> Self {
        let mut allowed_dirs = Vec::new();
        allowed_dirs.push("/tmp".to_string());
        allowed_dirs.push("/var/tmp".to_string());
        allowed_dirs.push("/home/user/projectx".to_string());

        let mut allowed_commands = HashSet::new();
        allowed_commands.insert("echo".to_string());
        allowed_commands.insert("date".to_string());
        allowed_commands.insert("hostname".to_string());
        allowed_commands.insert("pwd".to_string());
        allowed_commands.insert("ls".to_string());
        allowed_commands.insert("cat".to_string());
        allowed_commands.insert("head".to_string());
        allowed_commands.insert("tail".to_string());

        Self {
            allowed_dirs,
            allowed_commands,
            max_output_size: 1024 * 1024, // 1MB
            timeout_secs: 30,
            max_command_length: 256,
        }
    }
}

/// Safe Shell Executor
pub struct SafeShell {
    config: ShellConfig,
    /// Shell injection patterns
    injection_patterns: Vec<Regex>,
    /// Blocked patterns
    blocked_patterns: Vec<Regex>,
}

impl SafeShell {
    /// Create new executor
    pub fn new() -> Self {
        Self::with_config(ShellConfig::default())
    }

    /// Create with custom config
    pub fn with_config(config: ShellConfig) -> Self {
        // Shell injection patterns
        let injection_patterns = vec![
            Regex::new(r"\$\([^)]+\)").unwrap(),           // Command substitution $()
            Regex::new(r"`[^`]+`").unwrap(),               // Backtick substitution
            Regex::new(r"\|\s*bash").unwrap(),             // Pipe to bash
            Regex::new(r"\|\s*sh").unwrap(),                // Pipe to sh
            Regex::new(r"&&").unwrap(),                     // Command chaining
            Regex::new(r"\|\|").unwrap(),                   // OR chaining
            Regex::new(r";\s*rm\s").unwrap(),              // Semicolon + rm
            Regex::new(r";\s*wget\s").unwrap(),            // Semicolon + wget
            Regex::new(r";\s*curl\s").unwrap(),            // Semicolon + curl
            Regex::new(r";\s*nc\s").unwrap(),              // Semicolon + netcat
            Regex::new(r"\bexec\s").unwrap(),              // exec command
            Regex::new(r"\beval\s").unwrap(),              // eval command
            Regex::new(r"\bsource\s").unwrap(),            // source command
            Regex::new(r"\.\s*/").unwrap(),                // Execute from path
        ];

        // Blocked path patterns
        let blocked_patterns = vec![
            Regex::new(r"^/etc").unwrap(),      // System config
            Regex::new(r"^/proc").unwrap(),     // Process info
            Regex::new(r"^/sys").unwrap(),      // Kernel info
            Regex::new(r"^/root").unwrap(),    // Root home
            Regex::new(r"^/var/log").unwrap(), // Logs
            Regex::new(r"^/home/[^/]+/\.ssh").unwrap(), // SSH keys
            Regex::new(r"^/dev").unwrap(),     // Devices
        ];

        Self {
            config,
            injection_patterns,
            blocked_patterns,
        }
    }

    /// Check for shell injection
    fn check_injection(&self, command: &str) -> Result<(), SafeShellError> {
        for pattern in &self.injection_patterns {
            if pattern.is_match(command) {
                return Err(SafeShellError::InjectionDetected(
                    format!("Pattern matched: {}", pattern),
                ));
            }
        }
        Ok(())
    }

    /// Check for blocked paths
    fn check_path(&self, path: &str) -> Result<(), SafeShellError> {
        // Check blocked patterns
        for pattern in &self.blocked_patterns {
            if pattern.is_match(path) {
                return Err(SafeShellError::DirectoryNotAllowed(
                    format!("Blocked path: {}", path),
                ));
            }
        }

        // Check allowed directories
        let canonical = Path::new(path);
        let is_allowed = self.config.allowed_dirs.iter().any(|dir| {
            canonical.starts_with(dir) || canonical.to_string_lossy().starts_with(dir)
        });

        if !is_allowed {
            return Err(SafeShellError::DirectoryNotAllowed(format!(
                "Path not in allowed list: {}",
                path
            )));
        }

        Ok(())
    }

    /// Extract command from input
    fn extract_command(&self, input: &str) -> Result<String, SafeShellError> {
        let trimmed = input.trim();
        
        // Check length
        if trimmed.len() > self.config.max_command_length {
            return Err(SafeShellError::ExecutionFailed(
                "Command too long".to_string(),
            ));
        }

        // Extract base command (first word)
        let parts: Vec<&str> = trimmed.split_whitespace().collect();
        if parts.is_empty() {
            return Err(SafeShellError::ExecutionFailed(
                "Empty command".to_string(),
            ));
        }

        let cmd = parts[0];

        // Check whitelist
        if !self.config.allowed_commands.contains(cmd) {
            return Err(SafeShellError::CommandNotAllowed(format!(
                "Command not in whitelist: {}",
                cmd
            )));
        }

        Ok(trimmed.to_string())
    }

    /// Execute a safe shell command
    pub fn execute(&self, command: &str) -> Result<ShellResult, SafeShellError> {
        let start = Instant::now();

        // 1. Check for injection
        self.check_injection(command)?;

        // 2. Extract and validate command
        let validated_cmd = self.extract_command(command)?;

        // 3. Check for dangerous paths in arguments
        for arg in validated_cmd.split_whitespace() {
            if arg.starts_with('/') {
                self.check_path(arg)?;
            }
        }

        // 4. Execute with timeout
        let output = Command::new("sh")
            .args(["-c", &validated_cmd])
            .stdout(Stdio::piped())
            .stderr(Stdio::piped())
            .output();

        let duration = start.elapsed();

        match output {
            Ok(output) => {
                let stdout = String::from_utf8_lossy(&output.stdout).to_string();
                let stderr = String::from_utf8_lossy(&output.stderr).to_string();

                // Check output size
                if stdout.len() + stderr.len() > self.config.max_output_size {
                    return Err(SafeShellError::OutputTooLarge(format!(
                        "Output exceeds {} bytes",
                        self.config.max_output_size
                    )));
                }

                Ok(ShellResult {
                    success: output.status.success(),
                    stdout,
                    stderr,
                    exit_code: output.status.code().unwrap_or(-1),
                    duration_ms: duration.as_millis() as u64,
                })
            }
            Err(e) => Err(SafeShellError::ExecutionFailed(e.to_string())),
        }
    }

    /// Execute with timeout
    pub fn execute_with_timeout(&self, command: &str) -> Result<ShellResult, SafeShellError> {
        let timeout = Duration::from_secs(self.config.timeout_secs);

        // Use std::thread::spawn for timeout (simpler than async)
        let command = command.to_string();
        
        let handle = std::thread::spawn(move || {
            let shell = SafeShell::new();
            shell.execute(&command)
        });

        match handle.join() {
            Ok(result) => result,
            Err(_) => Err(SafeShellError::Timeout("Thread panicked".to_string())),
        }
    }

    /// Add allowed directory
    pub fn add_allowed_dir(&mut self, dir: &str) {
        self.config.allowed_dirs.push(dir.to_string());
    }

    /// Add allowed command
    pub fn add_allowed_command(&mut self, cmd: &str) {
        self.config.allowed_commands.insert(cmd.to_string());
    }
}

impl Default for SafeShell {
    fn default() -> Self {
        Self::new()
    }
}

// Global executor
use once_cell::sync::Lazy;
use std::sync::RwLock;

static SAFE_SHELL: Lazy<RwLock<SafeShell>> = Lazy::new(|| RwLock::new(SafeShell::new()));

/// Initialize global shell
pub fn init() {
    let _ = SAFE_SHELL.read();
}

/// Execute command
pub fn execute(command: &str) -> Result<ShellResult, SafeShellError> {
    SAFE_SHELL.read().unwrap().execute(command)
}

/// Execute with timeout
pub fn execute_with_timeout(command: &str) -> Result<ShellResult, SafeShellError> {
    SAFE_SHELL.read().unwrap().execute_with_timeout(command)
}

/// Add allowed directory
pub fn add_allowed_dir(dir: &str) {
    SAFE_SHELL.write().unwrap().add_allowed_dir(dir);
}

/// Add allowed command
pub fn add_allowed_command(cmd: &str) {
    SAFE_SHELL.write().unwrap().add_allowed_command(cmd);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_allowed_command() {
        let shell = SafeShell::new();
        let result = shell.execute("echo hello").unwrap();
        assert!(result.success);
    }

    #[test]
    fn test_blocked_command() {
        let shell = SafeShell::new();
        let result = shell.execute("cat /etc/passwd");
        assert!(result.is_err());
    }

    #[test]
    fn test_injection_blocked() {
        let shell = SafeShell::new();
        let result = shell.execute("echo test; rm -rf /");
        assert!(result.is_err());
    }
}
